numpy/core/src/multiarray/item_selection.c

```diff
@@ -2093,13 +2093,17 @@ PyArray_CountNonzero(PyArrayObject *self)
         PyArray_PREPARE_TRIVIAL_ITERATION(self, count, data, stride);
 
         while (count--) {
-            if (nonzero(data, self)) {
+            npy_bool n = nonzero(data, self);
+            if (n == -1) {
```

Comment:
Well, correctness beats performance. If we are required by Python to check inside the loop, we may have to accept the slowdown (or find some workaround such as a special loop for Object dtypes).

Particularly in the case of Object types, I think the call to `PyObject_IsTrue` can enter back to python code, so I wouldn't want to leave stray error states.